src/Components/Server/src/Circuits/CircuitIdFactory.cs

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Security.Cryptography;
+using Microsoft.AspNetCore.DataProtection;
+using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Components.Server.Circuits
+{
+    // This is a singleton instance
+    // Generates strong criptographic ids for circuits that are protected with authenticated encryption.
+    internal class CircuitIdFactory
+    {
+        private readonly RandomNumberGenerator _generator = RandomNumberGenerator.Create();
+        private readonly IDataProtector _protector;
+
+        public CircuitIdFactory(IOptions<CircuitOptions> options)
+        {
+            _protector = options.Value.CircuitIdProtector;
+        }
+
+        // Generates a circuit id that is produced from a strong cryptographic random number generator
+        // we don't care about the underlying payload, other than its uniqueness and the fact that we
+        // authenticate encrypt it using data protection.
+        // For validation, the fact that we can unprotect the payload is guarantee enough.
+        public string CreateCircuitId()
+        {
+            var buffer = new byte[32];
+            try
+            {
+                _generator.GetBytes(buffer);
+                var payload = _protector.Protect(buffer);
+
+                return Base64UrlTextEncoder.Encode(payload);
+            }
+            finally
+            {
+                // Its good practice to clear secrets from memory as soon as we are done with them.
+                Array.Clear(buffer, 0, 32);
```

Comment:
^-^ There's nothing secret about this value. It's protected to prevent tampering.

Reply:
Fair enough, I just get on paranoid mode when I write this type of code. You are right in that we don't need to necessarily clear it.